### PR TITLE
[Operator Versioning][Test] Automate model generating process

### PR DESCRIPTION
**Pull request by cccclai**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70629

This change is to automate the process to generate the old models for testing upgraders. Developer will
1. Add a module in `caffe2/test/jit/fixtures_srcs/fixtures_src.py`
2. Register the module in `caffe2/test/jit/fixtures_srcs/generate_models.py`
3. Run `python test/jit/fixtures_src/generate_models.py`

The model will be dumped to `test/jit/fixtures`.

This script also ensure:
1. The output model operator version is as expected
2. The output model will include the changed operator

Example log:
```
(base) chenlai@chenlai-mp pytorch % python3 /Users/chenlai/pytorch/test/jit/fixtures_src/generate_models.py
TestVersionedDivTensorExampleV4() aten::div.Tensor
INFO:__main__:Processing TestVersionedDivTensorExampleV4
INFO:__main__:Generating model test_versioned_div_tensor_example_v4 and it's save to /Users/chenlai/pytorch/test/jit/fixtures/test_versioned_div_tensor_example_v4.ptl
```
The second time to run
```
(base) chenlai@chenlai-mp pytorch % python3 /Users/chenlai/pytorch/test/jit/fixtures_src/generate_models.py
TestVersionedDivTensorExampleV4() aten::div.Tensor
INFO:__main__:Processing TestVersionedDivTensorExampleV4
INFO:__main__:Model test_versioned_div_tensor_example_v4 already exists, skipping
```

Differential Revision: [D33410841](https://our.internmc.facebook.com/intern/diff/D33410841/)

Differential Revision: [D33410841](https://our.internmc.facebook.com/intern/diff/D33410841)

### CODE REVIEW

File: test/jit/fixtures_srcs/generate_models.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+an upgrader is needed. Here is the flow to properly land a BC-breaking operator change.
+
+1. Write an upgrader in caffe2/torch/jit/operator_upgraders.py file. The softly enforced naming format
+is <operator_name>_<operator_overload>_<start>_<end>. For example, the below example means that div.Tensor at
+version from 0 to 3 needs to be replaced by this upgrader.
+
+```
+div_Tensor_0_3 is added for a change of operator div in pr xxxxxxx.
+Create date: 12/02/2021
+Expire date: 06/02/2022
```
Comment (iseeyuan) [inline test/jit/fixtures_srcs/generate_models.py:30]:
```suggestion
# div_Tensor_0_3 is added for a change of operator div in pr xxxxxxx.
# Create date: 12/02/2021
# Expire date: 06/02/2022
```

File: test/jit/fixtures_srcs/generate_models.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+3. After rebuild PyTorch, run the following command and it will auto generate a change to
+fbcode/caffe2/torch/csrc/jit/mobile/upgrader_mobile.cpp
+
+```
+python pytorch/tools/codegen/operator_versions/gen_mobile_upgraders.py
+```
+
+4. Generate the test to cover upgrader.
+4.1 Switch the commit before the operator change, and add a module in
```
Comment (iseeyuan) [inline test/jit/fixtures_srcs/generate_models.py:55]: Could the test be added in the operator change commit?

**Review by iseeyuan — APPROVED** [review]
Looks nice! Left two minor comments. Please address them before landing it.

File: test/jit/fixtures_srcs/generate_models.py (hunk 2 above)
Comment (cccclai) [reply test/jit/fixtures_srcs/generate_models.py:55]: Yeah the test need to be added with the operator change commit. This script needs to run before the commit to do the operator change though. Let me update this part so it's clearer.

File: test/jit/fixtures_srcs/generate_models.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+"""
+This file is used to generate model for test operator change. Please refer to
+https://github.com/pytorch/rfcs/blob/master/RFC-0017-PyTorch-Operator-Versioning.md for more details.
+
+A systematic workflow to change operator is needed to ensure
+Backwards Compatibility (BC) / Forwards Compatibility (FC) for operator changes. For BC-breaking operator change,
+an upgrader is needed. Here is the flow to properly land a BC-breaking operator change.
+
+1. Write an upgrader in caffe2/torch/jit/operator_upgraders.py file. The softly enforced naming format
```
Comment (tugsbayasgalan) [inline test/jit/fixtures_srcs/generate_models.py:23]: Can this wait to be landed until I land the change where we move everything to cpp?

File: test/jit/fixtures_srcs/generate_models.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+"""
+
+"""
+A map of test modules and it's according changed operator
+key: test module
+value: changed operator
+"""
+ALL_MODULES = {
+    TestVersionedDivTensorExampleV4(): "aten::div.Tensor",
```
Comment (tugsbayasgalan) [inline test/jit/fixtures_srcs/generate_models.py:72]: We include other operators here right?

File: test/jit/fixtures_srcs/generate_models.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+2. The output model version is not the same as expected version. For example, test_versioned_div_tensor_example_v4
+is used to test an operator change aten::div.Tensor, and the operator version will be bumped to v5. This script is
+supposed to run before the operator change (before the commit to make the change). If the actual model version is v5,
+likely this script is running with the commit to make the change.
+
+3. The model already exists in `test/jit/fixtures`.
+
+"""
+def generate_models():
```
Comment (tugsbayasgalan) [inline test/jit/fixtures_srcs/generate_models.py:135]: this is very neat! Thank you for writing it :)

File: test/jit/fixtures_srcs/generate_models.py (hunk 3 above)
Comment (cccclai) [reply test/jit/fixtures_srcs/generate_models.py:23]: Yeah of course.

File: test/jit/fixtures_srcs/generate_models.py (hunk 4 above)
Comment (cccclai) [reply test/jit/fixtures_srcs/generate_models.py:72]: Hmm I think one module is supposed to cover operator...basically, for every operator change, one model will be generated to cover it.

File: test/jit/fixtures_srcs/generate_models.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
+import io
+import logging
+import sys
+import zipfile
+from pathlib import Path
+from typing import Set
+
+import torch
+from test.jit.fixtures_srcs.fixtures_src import TestVersionedDivTensorExampleV4
```
Comment (tugsbayasgalan) [inline test/jit/fixtures_srcs/generate_models.py:9]: should we just import everything from there? So that developers have one less thing to worry about

**Review by tugsbayasgalan — APPROVED** [review]

File: test/jit/fixtures_srcs/generate_models.py (hunk 6 above)
Comment (cccclai) [reply test/jit/fixtures_srcs/generate_models.py:9]: Good point!

**Pull request closed without merging**